### PR TITLE
[COCO] Enable cutting logic

### PR DESCRIPTION
**Pull request by aalborov**

### CODE REVIEW

File: .idea/.gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+
```
Comment (demid5111) [inline .idea/.gitignore]: rm .idea folder + add to gitignore

File: cut_coco.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+                        required=True,
+                        help='Full path to the source archive')
+    parser.add_argument('--source_annotations_archive_dir',
+                        type=str,
+                        required=True,
+                        help='Full path to the source archive')
+    parser.add_argument('--first_image',
+                        type=int,
+                        required=True,
+                        help='Number of images in the output dataset')
```
Comment (demid5111) [inline cut_coco.py]: does not seem to correspond with the parameter

Comment (demid5111) [reply cut_coco.py]: required?

File: cut_coco.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    parser.add_argument('--output_archive_dir',
+                        type=str,
+                        required=True,
+                        help='Full path to the output archive (without the name of the archive)')
+    return parser
+
+
+def unarchive(source_archive_dir, output_folder_dir):
```
Comment (demid5111) [inline cut_coco.py]: DRY - Do not Repeat Yourself

File: cut_coco.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    new_image_filenames = []
+    for image in image_filenames:
+        new_image_filenames.append(os.path.join(images_folder_dir, image))
+
+    files_to_archive = new_image_filenames
+    files_to_archive.append(annotations_folder)
+    return files_to_archive
+
+
+def archive(new_file_names, source_path, output_archive_name, output_folder_dir):
```
Comment (demid5111) [inline cut_coco.py]: 🥇

File: cut_coco.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            else:
+                relative_path = '/{}'.format(file_name[file_name.find(rel_path_finder):])
+            tar.add(os.path.join(output_folder_dir, file_name), arcname=relative_path)
+
+
+def clean_up(output_folder_dir):
+    shutil.rmtree(output_folder_dir)
+
+if __name__ == '__main__':
+    args = parser().parse_args()
```
Comment (demid5111) [inline cut_coco.py]: move to the common script

File: cut_coco.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+
+def cut_coco(first_image, output_size, output_folder_dir):
+    folders = os.listdir(output_folder_dir)
+
+    annotation_files = []
+    if len(folders) != 2:
+        sys.exit('Incorrect dataset type')
+    for folder in folders:
+        if 'trainval' and 'annotations' in folder:
```
Comment (demid5111) [inline cut_coco.py]: use regex

File: cut_coco.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+def unarchive(source_archive_dir, output_folder_dir):
+    shutil.unpack_archive(source_archive_dir, output_folder_dir)
+
+
+def cut_coco(first_image, output_size, output_folder_dir):
+    folders = os.listdir(output_folder_dir)
+
+    annotation_files = []
+    if len(folders) != 2:
```
Comment (demid5111) [inline cut_coco.py]: magic constant

File: cut_coco.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    for folder in folders:
+        if 'trainval' and 'annotations' in folder:
+            annotations_folder = os.path.join(output_folder_dir, folder)
+            annotations = os.listdir(annotations_folder)
+        else:
+            images_folder_dir = os.path.join(output_folder_dir, folder)
+            images_folder = os.listdir(images_folder_dir)
+    for file_name in annotations:
+        file_name_l = file_name.lower()
+        if file_name_l.endswith('.json') and 'instances_val' in file_name_l:
```
Comment (demid5111) [inline cut_coco.py]: find by regex with system utilities

File: cut_coco.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+            file_dir = os.path.join(output_folder_dir, annotations_folder, file_name)
+            annotation_files.append(file_dir)
+            if len(annotation_files) > 1:
```
Comment (demid5111) [inline cut_coco.py]: no need in the check the loop

File: cut_coco.py [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+                sys.exit('Incorrect dataset format. 1')
```
Comment (demid5111) [inline cut_coco.py]: 👍

File: cut_coco.py [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+            else:
+                annotation_name = file_name
+                annotation_dir = file_dir
+    if len(images_folder) == 0:
```
Comment (demid5111) [inline cut_coco.py]: interesting, try something more python-ish

File: cut_coco.py [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+        sys.exit('Incorrect dataset format. 2')
+
+    if first_image >= len(images_folder) - output_size:
```
Comment (demid5111) [inline cut_coco.py]: move up, need to raise error earlier

File: cut_coco.py [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+    json_data['annotations'] = cut_annotations
+
+    with open(annotation_name, 'w') as outfile:
+        json.dump(json_data, outfile)
+
+    new_image_filenames = []
+    for image in image_filenames:
+        new_image_filenames.append(os.path.join(images_folder_dir, image))
+
+    files_to_archive = new_image_filenames
```
Comment (demid5111) [inline cut_coco.py]: it does not solve immutability pattern

**Review by demid5111 — CHANGES_REQUESTED** [review]

File: .idea/.gitignore (hunk 1 above)
Comment (aalborov) [reply .idea/.gitignore]: done as a separate PR

File: cut_coco.py [hunk 14]
```diff
@@ -45,22 +46,22 @@ def cut_coco(output_size, output_folder_dir, first_image):
...
-        if file_name_l.endswith('.json') and 'instances_val' in file_name_l:
+        if re.match('instances_val\d{2,4}.json', file_name_l):
             file_dir = os.path.join(output_folder_dir, annotations_folder, file_name)
             annotation_files.append(file_dir)
-            if len(annotation_files) > 1:
-                sys.exit('Incorrect dataset format.')
-            else:
-                annotation_name = file_name
-                annotation_dir = file_dir
+        if len(annotation_files) > 1:
```
Comment (aalborov) [inline cut_coco.py]: indent

File: cut_coco.py [hunk 15]
```diff
@@ -45,22 +46,22 @@ def cut_coco(output_size, output_folder_dir, first_image):
     if len(folders) != num_of_folders:
         sys.exit('Incorrect dataset type')
     for folder in folders:
-        if 'trainval' and 'annotations' in folder:
+        if re.match('annotations_trainval\d{2,4}$', folder):
```
Comment (aalborov) [inline cut_coco.py]: no loop

**Comment by aalborov** [comment]
[VOC_sample.zip](https://github.com/aalborov/cut_dataset/files/3886963/VOC_sample.zip)

**Comment by aalborov** [comment]
[sample_imagenet.zip](https://github.com/aalborov/cut_dataset/files/3890377/sample_imagenet.zip)
[val10.zip](https://github.com/aalborov/cut_dataset/files/3890379/val10.zip)
[annotations_trainval10.zip](https://github.com/aalborov/cut_dataset/files/3890380/annotations_trainval10.zip)

**Comment by aalborov** [comment]
[VOC_sample.zip](https://github.com/aalborov/cut_dataset/files/3890387/VOC_sample.zip)

File: cut_dataset.py [hunk 16]
```diff
@@ -158,24 +214,35 @@ def clean_up(output_folder_dir):
...
-    output_folder_dir = os.path.join(args.output_archive_dir, 'temp')
+    output_folder_dir = os.path.join(args.output_archive_dir, 'subset_folder')
     output_archive_name = '{}_subset_{}_{}'.format(args.dataset_type, args.first_image, args.first_image + args.output_size)
-    unarchive(args.source_archive_dir, output_folder_dir)
 
-    if is_imagenet(args.dataset_type):
-        imagenet_data = cut_imagenet(args.output_size, output_folder_dir, args.first_image)
-        new_file_names = imagenet_data[0]
-        rel_path_finder = imagenet_data[1]
+    if not is_coco(args.dataset_type):
```
Comment (aalborov) [inline cut_dataset.py]: if isvoc and not args.sourceDir or  :
  exit

**Comment by aalborov** [comment]
[sample_voc.zip](https://github.com/aalborov/cut_dataset/files/3895549/sample_voc.zip)

**Comment by aalborov** [comment]
[sample_voc.zip](https://github.com/aalborov/cut_dataset/files/3895560/sample_voc.zip)

**Pull request merged**